File: python/Baekjoon/sam/bfs.py
from collections import deque

def bfs(graph, start, visited):
    # 큐 생성, 현재 노드 방문 처리
    queue = deque([start])
    visited[start] = True

    # 큐가 빌 때까지 반복
    while queue:
        # 큐에서 원소 하나 뽑아서 출력
        v = queue.popleft()
        print(v, end=" ")

        # 아직 방문하지 않은 인접 노드를 큐에 넣음
        for node in graph[v]:
            if not (visited[node]):
                queue.append(node)
                visited[node] = True


graph = [
    [],
    [2, 3, 8],
    [1, 7],
    [1, 4, 5],
    [3, 5],
    [3, 4],
    [7],
    [2, 6, 8],
    [1, 7]
]

visited = [False] * 9
bfs(graph, 1, visited)

File: c++/basic/step2/bfs.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> adj[100]; int visited[100];

void BFS(int start) {
    queue<int> q;
    visited[start] = 1;
    q.push(start);

    while(q.size()) {
        int here = q.front(); q.pop();
        for(int there:adj[here]) {
            if(visited[there]) continue;
            visited[there] = visited[here] + 1;
            q.push(there);
        }
    }
}

int main() {
    adj[10].push_back(12);
    adj[10].push_back(14);
    adj[10].push_back(16);
    adj[12].push_back(18);
    adj[12].push_back(20);
    adj[20].push_back(22);
    adj[20].push_back(24);

    BFS(10);

    int nodeList[] = {10, 12, 14, 16, 18, 20, 22, 24};
    for(int i : nodeList){
        cout << i << " : " << visited[i] << '\n';
    }

    return 0;
}